<a href="https://colab.research.google.com/github/Ryan-M-Smith/CS315/blob/main/InClass/quicksort.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [28]:
# Template for testing Quick-Sort on small lists, outputing each intermediate step

In [29]:
import random
import time
import numpy as np
from typing import Any

In [30]:
# Partition provided as a convenience

def partition(arr: list[Any], p: int, r: int) -> int:
    x = arr[r]
    i = p - 1
    
    for j in range(p,r):
        if arr[j] <= x:
            i += 1
            arr[i], arr[j] = arr[j], arr[i]    # Swap arr[i] and arr[j]
            
    arr[i+1], arr[r] = arr[r], arr[i+1]        # Swap arr[i+1] and arr[r]

    return i + 1

In [31]:
def quicksort(arr: list[Any], p: int, r: int, *, print_flag: bool = False) -> None:
    if p < r:
        q = partition(arr, p, r)
        
        if print_flag:
            print(arr)
        
        quicksort(arr, p, q - 1, print_flag=print_flag)
        quicksort(arr, q + 1 , r, print_flag=print_flag)

In [32]:
n = 8
arr = list(range(n))
random.shuffle(arr)

print(arr)
quicksort(arr, 0, n - 1, print_flag=True)
print(arr)

[4, 5, 0, 3, 6, 1, 2, 7]
[4, 5, 0, 3, 6, 1, 2, 7]
[0, 1, 2, 3, 6, 5, 4, 7]
[0, 1, 2, 3, 6, 5, 4, 7]
[0, 1, 2, 3, 4, 5, 6, 7]
[0, 1, 2, 3, 4, 5, 6, 7]
[0, 1, 2, 3, 4, 5, 6, 7]


In [33]:
#
# HW04 - time Quicksort on random data for a variety of array sizes.
#

rng = np.random.default_rng()
TEST_RUNS = 10
sizes = [10, 50, 100, 1000, 2500, 7500, 10000, 12500, 17500, 20000]

for i in range(TEST_RUNS):
        n = sizes[i]
        arr = rng.integers(low=0, high=100, size=n).tolist()

        start = time.perf_counter()
        quicksort(arr, 0, n - 1)
        end = time.perf_counter()

        print(f"n = {n:<8,} time = {(end - start) * 1_000:.3f} ms")

n = 10       time = 0.006 ms
n = 50       time = 0.022 ms
n = 100      time = 0.900 ms
n = 1,000    time = 1.816 ms
n = 2,500    time = 5.161 ms
n = 7,500    time = 26.336 ms
n = 10,000   time = 45.511 ms
n = 12,500   time = 63.852 ms
n = 17,500   time = 118.799 ms
n = 20,000   time = 153.088 ms


## Results

Because Merge Sort has an expected runtime of $\Theta\left(n^2\right)$, we will calculate all ratios with a denominator of $n^2$.

| $n$    | $T(n)$      | $\dfrac{T(n)}{n \lg n}$      |
| ------ | ----------- | ---------------------------- |
| 10     | 0.006       | 0.000180                     |
| 50     | 0.022       | 0.000538                     |
| 100    | 0.900       | 0.01354                      |
| 1,000  | 1.816       | 0.000182                     |
| 2,500  | 5.161       | 0.000183                     |
| 7,500  | 26.336      | 0.000272                     |
| 10,000 | 45.511      | 0.000271                     |
| 12,500 | 63.852      | 0.000374                     |
| 17,500 | 118.799     | 0.000540                     |
| 20,000 | 153.088     | 0.000464                     |



Looking at our ratios, as $n$ gets larger, the results tend to stabilize around a factor of $1.0 \times 10^{-4}$, showing that the runtime tends to stabilize
around a ratio of $n \lg n$. This strongly suggests that the algorithm runs in $\Theta\left(n \lg n\right)$ time.


In [34]:
#
# HW04 - time Quicksort on sorted data for a variety of array sizes.
#

# This code will run into recursion depth issues for large n because quicksort
# will be running in a worst-case scenario.
import sys
sys.setrecursionlimit(25000)

for i in range(TEST_RUNS):
    n = sizes[i]
    arr = list(range(n))

    start = time.perf_counter()
    quicksort(arr, 0, n - 1)
    end = time.perf_counter()

    print(f"n = {n:<8,} time = {(end - start)*1_000:.3f} ms")

n = 10       time = 0.008 ms
n = 50       time = 0.079 ms
n = 100      time = 0.302 ms
n = 1,000    time = 30.955 ms
n = 2,500    time = 201.886 ms
n = 7,500    time = 1923.617 ms
n = 10,000   time = 3299.818 ms
n = 12,500   time = 5247.543 ms
n = 17,500   time = 9953.997 ms
n = 20,000   time = 13118.797 ms
